File: python/minecraft_bot/protocol/v763/packets/play/serverbound/use_entity.py
"""Packet `use_entity` (play/serverbound, id 0x10).

Player attacks or uses an entity. ``mouse``: 0=interact, 1=attack,
2=interact-at. For ``mouse == 2`` the packet carries hit-position
fields.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from typing import Optional

from minecraft_bot.codec import Reader, Writer, varint
from minecraft_bot.errors import ValueOutOfRange

PACKET_ID = 0x10


@dataclass(frozen=True, slots=True)
class UseEntity:
    target: int                          # varint, target entity id
    mouse: int                           # varint, 0/1/2
    x: Optional[float]                   # f32, only when mouse == 2
    y: Optional[float]
    z: Optional[float]
    hand: Optional[int]                  # varint, only when mouse == 0 or 2
    sneaking: bool


def _bool(reader: Reader, name: str) -> bool:
    b = reader.read(1)[0]
    if b not in (0, 1):
        raise ValueOutOfRange(name, b)
    return b == 1


def decode(reader: Reader) -> UseEntity:
    target = varint.read(reader)
    mouse = varint.read(reader)
    x: Optional[float] = None
    y: Optional[float] = None
    z: Optional[float] = None
    hand: Optional[int] = None
    if mouse == 2:
        x, y, z = struct.unpack(">fff", reader.read(12))
    if mouse == 0 or mouse == 2:
        hand = varint.read(reader)
    sneaking = _bool(reader, "use_entity.sneaking")
    return UseEntity(target=target, mouse=mouse, x=x, y=y, z=z, hand=hand, sneaking=sneaking)


def encode(packet: UseEntity, writer: Writer) -> None:
    varint.write(packet.target, writer)
    varint.write(packet.mouse, writer)
    if packet.mouse == 2:
        if packet.x is None or packet.y is None or packet.z is None:
            raise ValueOutOfRange("use_entity.xyz", None)
        writer.write(struct.pack(">fff", packet.x, packet.y, packet.z))
    if packet.mouse in (0, 2):
        if packet.hand is None:
            raise ValueOutOfRange("use_entity.hand", None)
        varint.write(packet.hand, writer)
    writer.write(b"\x01" if packet.sneaking else b"\x00")

File: rust/src/protocol/v763/packets/play/serverbound/use_entity.rs
//! Packet `use_entity` (play/serverbound). Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot,
    string_codec as string, varint, varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ServerboundPacket;

pub const PACKET_ID: i32 = 0x10;

#[derive(Debug, Clone, PartialEq, Default)]
pub struct UseEntity {
    pub target: i32,
    pub mouse: i32,
    pub x: Option<f32>,
    pub y: Option<f32>,
    pub z: Option<f32>,
    pub hand: Option<i32>,
    pub sneaking: bool,
}

impl UseEntity {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let target = varint::read(reader)?;
        let mouse = varint::read(reader)?;
        let (mut x, mut y, mut z) = (None, None, None);
        let mut hand = None;
        if mouse == 2 {
            let b = reader.read_exact(12)?;
            x = Some(f32::from_be_bytes([b[0],b[1],b[2],b[3]]));
            y = Some(f32::from_be_bytes([b[4],b[5],b[6],b[7]]));
            z = Some(f32::from_be_bytes([b[8],b[9],b[10],b[11]]));
        }
        if mouse == 0 || mouse == 2 { hand = Some(varint::read(reader)?); }
        let bn = reader.read_exact(1)?[0];
        if bn > 1 { return Err(ProtocolError::DecodeError(format!("sneaking: {}", bn))); }
        Ok(Self { target, mouse, x, y, z, hand, sneaking: bn != 0 })
    }
}

impl ServerboundPacket for UseEntity {
    fn state(&self) -> ConnectionState { ConnectionState::Play }
    fn packet_id(&self) -> i32 { PACKET_ID }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        varint::write(self.target, writer)?;
        varint::write(self.mouse, writer)?;
        if let (Some(x), Some(y), Some(z)) = (self.x, self.y, self.z) {
            writer.write_all(&x.to_be_bytes())?;
            writer.write_all(&y.to_be_bytes())?;
            writer.write_all(&z.to_be_bytes())?;
        }
        if let Some(h) = self.hand { varint::write(h, writer)?; }
        writer.write_all(&[if self.sneaking { 1 } else { 0 }])?;
        Ok(())
    }
}
